File: backend/src/main/java/ESP32/Emulator/command/CommandHandler.java
package ESP32.Emulator.command;

import ESP32.Emulator.actuator.Led;
import ESP32.Emulator.device.Device;
import ESP32.Emulator.device.DeviceRegistry;
import ESP32.Emulator.device.Esp32;

public class CommandHandler {
    private final DeviceRegistry registry;


    public CommandHandler(DeviceRegistry registry) {
        this.registry = registry;
    }

    public void execute(Command command) {
        Device device = registry.get(command.getDeviceId());

        if (device == null) {
            throw new IllegalArgumentException("Unknown device: " + command.getDeviceId());
        }
        command.execute(device);
    }
}

File: device/src/device/command/CommandHandler.cpp
#include <Arduino.h>

#include "device/command/CommandHandler.h"
#include "device/LedDevice.h"


CommandHandler::CommandHandler(DeviceRegistry* registry)
{
    this->registry = registry;
}


void CommandHandler::handle(Command command)
{
    Serial.println("Command received");
    Serial.print("Device: ");
    Serial.println(command.deviceId);

    Device* device = registry->find(
        command.deviceId.c_str()
    );


    if(device == nullptr)
    {
        Serial.println("Device not found");
        return;
    }


    if(command.action == "TURN_ON")
    {
        LedDevice* led = (LedDevice*) device;

        led->turnOn();

        Serial.println("LED ON");

        led->printState();
    }


    if(command.action == "TURN_OFF")
    {
        LedDevice* led = (LedDevice*) device;

        led->turnOff();

        Serial.println("LED OFF");

        led->printState();
    }
}